lib/mayaUsd/utils/targetLayer.cpp

```diff
@@ -28,6 +30,54 @@
 #include <maya/MObject.h>
 #include <maya/MString.h>
 
+namespace {
+
+const char kTargetLayerAttrName[] = "usdStageTargetLayer";
+const char kTargetLayerPrimPathAttrName[] = "usdStageTargetLayerPrimPath";
+
+/// \brief Find prim node that matches given layer id
+PXR_NS::PcpNodeRef
+findPrimNode(const PXR_NS::PcpNodeRef& primNode, const std::string& targetLayerId)
+{
+    // We need to store and check current depth level first (BFS), this is to
+    // preserve the opinions from strong to week
+    std::vector<PXR_NS::PcpNodeRef> childNodes;
+    TF_FOR_ALL(child, primNode.GetChildrenRange())
+    {
+        // The prim node can't have a variant.
+        if (child->GetPath().ContainsPrimVariantSelection()) {
+            continue;
+        }
+        // Confirm that the primNode is a direct contributor to the root prim.
+        bool result = child->IsRootNode();
+        // Confirm we are looking at one of the composition arcs that could lead us to a prim
+        // instance and not some other contributor like a specializes or such.
+        if (!result) {
+            auto arcType = child->GetArcType();
+            result = (arcType == PcpArcTypeReference) || (arcType == PcpArcTypePayload)
+                || (arcType == PcpArcTypeVariant);
+        }
+
+        if (result) {
+            if (child->GetLayerStack()->GetIdentifier().rootLayer->GetIdentifier()
+                == targetLayerId) {
+                return *child;
+            }
+            childNodes.emplace_back(*child);
+        }
+    }
+
+    for (const auto& child : childNodes) {
+        auto childPrimNode = findPrimNode(child, targetLayerId);
```

Comment:
The worry I would have is that in very large scene, this recursion could be very deep? I suppose the are not scene so deep it bust the call stack... but it could be a failure mode.

Reply:
I hope it's not. Our production has been using this logic for a few years for a few projects, users haven't reported any issue.